ext/pcre/php_pcre.c

```diff
@@ -2156,85 +2147,81 @@ static zend_string *php_pcre_replace_array(HashTable *regex, zval *replace, zend
 
 /* {{{ php_replace_in_subject
  */
-static zend_always_inline zend_string *php_replace_in_subject(zval *regex, zval *replace, zval *subject, size_t limit, size_t *replace_count)
+static zend_always_inline zend_string *php_replace_in_subject(
+	zend_string *regex_str, HashTable *regex_ht,
+	zend_string *replace_str, HashTable *replace_ht,
+	zend_string *subject, size_t limit, size_t *replace_count)
 {
 	zend_string *result;
-	zend_string *subject_str = zval_get_string(subject);
-
-	if (Z_TYPE_P(regex) != IS_ARRAY) {
-		result = php_pcre_replace(Z_STR_P(regex),
-								  subject_str,
-								  ZSTR_VAL(subject_str),
-								  ZSTR_LEN(subject_str),
-								  Z_STR_P(replace),
-								  limit,
-								  replace_count);
-		zend_string_release_ex(subject_str, 0);
+
+	if (regex_str) {
+		ZEND_ASSERT(replace_str != NULL);
+		result = php_pcre_replace(regex_str, subject, ZSTR_VAL(subject), ZSTR_LEN(subject),
+			replace_str, limit, replace_count);
 	} else {
-		result = php_pcre_replace_array(Z_ARRVAL_P(regex),
-										replace,
-										subject_str,
-										limit,
-										replace_count);
+		zend_string *subject_str = zend_string_copy(subject);
+		ZEND_ASSERT(regex_ht != NULL);
+		result = php_pcre_replace_array(regex_ht, replace_str, replace_ht, subject_str,
+			limit, replace_count);
```

Comment:
The string copy seems to be necessary here as otherwise I get a Segfault in ``preg_replace2.phpt`` which I'm not exactly sure *why* happens.

Reply:
It is indeed necessary in this case, but I think it would make a great deal more sense if you sunk it into php_pcre_replace_array itself. Right now it's asymmetric, in that the caller needs to copy, but replace_array will release.

Reply:
Seems indeed reasonable, will do that and push the update whit a git squash.